clang/lib/Sema/SemaExpr.cpp

```diff
@@ -9150,6 +9150,29 @@ Sema::CheckAssignmentConstraints(QualType LHSType, ExprResult &RHS,
   LHSType = Context.getCanonicalType(LHSType).getUnqualifiedType();
   RHSType = Context.getCanonicalType(RHSType).getUnqualifiedType();
 
+  // __builtin_counted_by_ref cannot be assigned to a variable, used in
```

Comment:
I wonder if this is too restrictive? Consider:
```
  #define alloc(P, FAM, COUNT) ({ \
     typeof(P) __p = NULL;                                        \
     __p = malloc(MAX(sizeof(*__p),                               \
                      sizeof(*__p) + sizeof(*__p->FAM) * COUNT)); \
     if (auto *ptr = __builtin_counted_by_ref(__p->FAM))          \
       *ptr = COUNT;                                              \
     __p;                                                         \
  })
```
which seems eminently reasonable to write so as to avoid repeated calls to `__builtin_counted_by_ref()`. And if we think that's reasonable, what about:
```
void __check(const void *ptr) __attribute__((diagnose_if(ptr == 0, "error", "you did a bad!"))) {}
...
auto *ptr = __builtin_counted_by_ref(__p->FAM); \
__check(ptr);                                   \
*ptr = COUNT;                                   \
```
and so on.

I worry we're setting up the constraints tightly enough that this has to be used *just right*, which seems like a pretty sharp edge.

Reply:
This is a strict requirement for Apple's `-fbounds-safety` checking. The `counted_by` counter cannot have its address taken. It's done to ensure that it and the FAM are always safely modified together.

Reply:
See the RFC for more details:

https://discourse.llvm.org/t/rfc-introducing-new-clang-builtin-builtin-get-counted-by/80836/

Reply:
Thanks! I don't insist on a change, but I still assert that this is too restrictive and has a rather sharp edge. We could do dataflow analysis to ensure that the counter and the FAM are modified together, but I realize that has tradeoffs too.

Reply:
@AaronBallman I hear you. `-fbounds-safety` is intentionally restrictive to provide strong guarantees and to avoid the need for sophisticated analysis to enforce the guarantees. Once the pointer is passed to another pointer variable, or a function parameter, it becomes really hard to keep track of what is pointing to the count and make sure the count is updated safely:

```C
bool foo(int *ptr) {
  *ptr = 100; // was it pointer to count or not?
}

auto *ptr = __builtin_counted_by_ref(__p->FAM);
foo(ptr); // 
```

We can relax the rule later a bit depending on the use cases to the extent it's still safe (e.g., allowing it to be passed to `const` attributed function), but it's much easier to relax things afterwards than trying to add more necessary restrictions later on.

Reply:
> Once the pointer is passed to another pointer variable, or a function parameter, it becomes really hard to keep track of what is pointing to the count and make sure the count is updated safely:

Yeah, I wasn't thinking we should relax the restrictions entirely; it's basically impossible to track once you pass the pointer to another function, for example. However, when used within the function calling `__builtin_counted_by_ref`, we have full visibility into what's going on. The two examples I posted above are things I would have a hard time telling someone were true positives to diagnose as a misuse of the API.

> We can relax the rule later a bit depending on the use cases to the extent it's still safe (e.g., allowing it to be passed to const attributed function), but it's much easier to relax things afterwards than trying to add more necessary restrictions later on.

100% agreed, and I don't insist on a change here, just some feedback to keep in mind for future possible directions.